File: solutions/typescript/dnd-character/1/dnd-character.ts
export class DnDCharacter {
 // This character has, among other things, six abilities; strength, dexterity, constitution, intelligence, wisdom and charisma.

  strength: number = DnDCharacter.generateAbilityScore();
  dexterity: number = DnDCharacter.generateAbilityScore();
  constitution: number = DnDCharacter.generateAbilityScore();
  intelligence: number = DnDCharacter.generateAbilityScore();
  wisdom: number = DnDCharacter.generateAbilityScore();
  charisma: number = DnDCharacter.generateAbilityScore();
  hitpoints: number = 10 + DnDCharacter.getModifierFor(this.constitution);

  public static generateAbilityScore(): number {
    const rolls = Array.from({ length: 4 }, () => Math.floor(Math.random() * 6) + 1);
    const lowestRoll = Math.min(...rolls);
    for (let i = 0; i < rolls.length; i++) {
        if (rolls[i] === lowestRoll) {
            rolls.splice(i, 1);
            break;
        }
    }

    var sum : number = 0;
    for (let i = 0; i < rolls.length; i++) {
        sum = sum + rolls[i];
    }

    return sum;
  }

  public static getModifierFor(abilityValue: number): number {
    return Math.floor((abilityValue - 10) / 2);
  }
}

File: solutions/c/dnd-character/1/dnd_character.c
#include "dnd_character.h"
#include <math.h>
#include <stdlib.h>
#include <time.h>

int modifier(int score) { return floor((score - 10) / 2.0); }

int ability(void) {
  int rolls[4];
  int minimum = 7;
  int total = 0;
  for (int i = 0; i < 4; i++) {
    rolls[i] = (rand() % 6 + 1);
    if (rolls[i] < minimum) {
      minimum = rolls[i];
    }
    total += rolls[i];
  }
  return total - minimum;
}

dnd_character_t make_dnd_character(void) {
  srand(time(NULL));
  dnd_character_t character;

  character.charisma = ability();
  character.constitution = ability();
  character.dexterity = ability();
  character.intelligence = ability();
  character.strength = ability();
  character.wisdom = ability();

  character.hitpoints = 10 + modifier(character.constitution);

  return character;
}
